File: ConvexHull.py
import matplotlib.pyplot as plt

def cross_product(o, a, b):
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])

def convex_hull(points):
    points = sorted(points)
    
    lower = []
    for p in points:
        while len(lower) >= 2 and cross_product(lower[-2], lower[-1], p) <= 0:
            lower.pop()
        lower.append(p)
    
    upper = []
    for p in reversed(points):
        while len(upper) >= 2 and cross_product(upper[-2], upper[-1], p) <= 0:
            upper.pop()
        upper.append(p)
    
    return lower[:-1] + upper[:-1]

def main():
    n = int(input("Enter the number of points: "))
    points = []
    for i in range(n):
        x, y = map(int, input(f"Enter coordinates of point {i+1} (x y): ").split())
        points.append((x, y))
    
    hull = convex_hull(points)
    
    print("Convex Hull:")
    for x, y in hull:
        print(f"({x}, {y})")
    
    # Visualization
    x_vals, y_vals = zip(*points)
    plt.scatter(x_vals, y_vals, label="Input Points")
    hull.append(hull[0])  # Close the hull
    hx_vals, hy_vals = zip(*hull)
    plt.plot(hx_vals, hy_vals, "r--", label="Convex Hull")
    plt.legend()
    plt.show()

if __name__ == "__main__":
    main()

File: Convexhull.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Point {
    int x, y;
    bool operator<(const Point& p) const {
        return x < p.x || (x == p.x && y < p.y);
    }
};

int crossProduct(Point o, Point a, Point b) {
    return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x);
}

vector<Point> convexHull(vector<Point>& points) {
    sort(points.begin(), points.end());
    
    vector<Point> lower, upper;
    
    for (const auto& p : points) {
        while (lower.size() >= 2 && crossProduct(lower[lower.size()-2], lower.back(), p) <= 0)
            lower.pop_back();
        lower.push_back(p);
    }
    
    for (int i = points.size() - 1; i >= 0; i--) {
        while (upper.size() >= 2 && crossProduct(upper[upper.size()-2], upper.back(), points[i]) <= 0)
            upper.pop_back();
        upper.push_back(points[i]);
    }
    
    lower.pop_back();
    upper.pop_back();
    
    lower.insert(lower.end(), upper.begin(), upper.end());
    return lower;
}

int main() {
    int n;
    cout << "Enter the number of points: ";
    cin >> n;
    vector<Point> points(n);
    
    for (int i = 0; i < n; i++) {
        cout << "Enter coordinates of point " << i + 1 << " (x y): ";
        cin >> points[i].x >> points[i].y;
    }
    
    vector<Point> hull = convexHull(points);
    
    cout << "Convex Hull:\n";
    for (const auto& p : hull) {
        cout << "(" << p.x << ", " << p.y << ")\n";
    }
    
    return 0;
}
